mpi4jax/collective_ops/allreduce.py

```diff
@@ -161,24 +161,115 @@ def mpi_allreduce_value_and_jvp(in_args, tan_args, op, comm):
     res = Allreduce(x, token=token, op=op, comm=comm)
 
     # Identify the correct adjoint
-    op = unpack_hashable(op)
+    op_ = unpack_hashable(op)
 
-    if op == _MPI.SUM:
-        jvp = (x_tan, token_tan)
+    if op_ == _MPI.SUM:
+        jvp = mpi_allreduce_p.bind(x_tan, token, op=op, comm=comm)
     else:
         raise NotImplementedError(
-            "The adjoint of allreduce for {} operation is not defined".format(op)
+            "The adjoint of allreduce for {} operation is not defined".format(op_)
         )
 
     return (res, jvp)
 
 
+def mpi_allreduce_transpose_rule(tan_args, *x_args, op, comm):
+    _, token = x_args
+    t, _ = tan_args
+    # Identify the correct adjoint
+    op_ = unpack_hashable(op)
+
+    if op_ == _MPI.SUM:
+        return mpi_allreduceT_p.bind(t, token, op=op, comm=comm)
+    else:
+        raise NotImplementedError(
+            "The linear transpose of allreduce for {} is not defined".format(op_)
+        )
+
+
 mpi_allreduce_p.multiple_results = True
 mpi_allreduce_p.def_impl(mpi_allreduce_impl)
 mpi_allreduce_p.def_abstract_eval(mpi_allreduce_abstract_eval)
 
 ad.primitive_jvps[mpi_allreduce_p] = mpi_allreduce_value_and_jvp
+ad.primitive_transposes[mpi_allreduce_p] = mpi_allreduce_transpose_rule
 
 # assign to the primitive the correct encoder
 xla.backend_specific_translations["cpu"][mpi_allreduce_p] = mpi_allreduce_xla_encode_cpu
 xla.backend_specific_translations["gpu"][mpi_allreduce_p] = mpi_allreduce_xla_encode_gpu
+
+
+mpi_allreduceT_p = Primitive("allreduceT_mpi")
+
+
+@enforce_types(
+    op=(_MPI.Op, HashableMPIType),
+    comm=(_MPI.Intracomm, HashableMPIType),
+    token=(type(None), xla.Token, core.Tracer),
+)
+def AllreduceT(x, op, comm=_MPI.COMM_WORLD, token=None):
+    if token is None:
+        token = create_token(x)
```

Comment:
A user will never call this, so this is only called internally.
as, internally, there is always a token, we don't need the default `token=None`, and will never be handling that case, right?